src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,182 @@
+import classnames from 'classnames';
+import type { ComponentChildren } from 'preact';
+import {
+  useCallback,
+  useContext,
+  useId,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'preact/hooks';
+
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+import { useClickAway } from '../../hooks/use-click-away';
+import { useKeyPress } from '../../hooks/use-key-press';
+import { MenuCollapseIcon, MenuExpandIcon } from '../icons';
+import Button from './Button';
+import SelectContext from './SelectContext';
+
+export type SelectProps<T> = {
+  selected: T;
+  onChange: (newValue: T) => void;
+  label: ComponentChildren;
+  children: ComponentChildren;
+  disabled?: boolean;
+};
+
+function SelectMain<T>({
+  label,
+  selected,
+  onChange,
+  children,
+  disabled,
+}: SelectProps<T>) {
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const closeDropdown = useCallback(() => setIsDropdownOpen(false), []);
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useRef<HTMLButtonElement | null>(null);
+  const buttonId = useId();
+  const [shouldDropUp, setShouldDropUp] = useState(false);
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeDropdown();
+    },
+    [closeDropdown, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the dropdown
+  useClickAway(wrapperRef, closeDropdown);
+  useKeyPress(['Escape'], closeDropdown);
+
+  // Vertical arrow key for options in the dropdown
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
```

Comment:
In the native select, keyboard arrows only loop over items within the popup, not the popup + current value display. Also in the native select, keyboard selection doesn't loop when you get to the top/bottom. I'd check the WAI-ARIA specs to see what they say about this.

Reply:
> in the native select, keyboard selection doesn't loop when you get to the top/bottom. I'd check the WAI-ARIA specs to see what they say about this.

This is the default behavior on `useArrowKeyNavigation`. I think it should be possible to add a new optional `loop` option to determine its behavior, but I would like to tackle that separately and not as part of this.

I will try to address the point in which the select itself is not part of the loop.

Reply:
> In the native select, keyboard arrows only loop over items within the popup, not the popup + current value display

If I exclude the toggle button, then `useArrowKeyNavigation` has a weird behavior that makes the second item to be the first one to be focused when pressing arrow down.

I have realized all existing use cases for `useArrowKeyNavigation` are working on lists of elements that are visible and mounted, so it's possible we need to tweak it for this to work the right way.

For now I will keep the toggle in the list of items to iterate, and create two separated issues to address the optional infinite loop and making sure the first list item is the one selected when pressing arrow down for the first time.